File: Lilly_Medchem_Rules.rb
#!/usr/bin/env ruby

# Run the rejection/demerit rules

ianhome=File.dirname($0)      # location for supporting files

require "#{ianhome}/ruby/iwcmdline.rb"
require "#{ianhome}/ruby/bindir.rb"

$expert = false

$default_lower_atom_count_cutoff = 7
$default_soft_upper_atom_count_cutoff = 25
$default_hard_upper_atom_count_cutoff = 40

def usage (rc)
  $stderr << "Runs the Lilly medchem rules\n"
  $stderr << " -c <n>           lower atom count cutoff (default #{$default_lower_atom_count_cutoff})\n" if $expert
  $stderr << " -Cs <natoms>     soft upper atom count cuttof (default #{$default_soft_upper_atom_count_cutoff})\n" if $expert
  $stderr << " -Ch <natoms>     hard upper atom count cuttof (default #{$default_hard_upper_atom_count_cutoff})\n" if $expert
  $stderr << " -C <natoms>      upper atom count limit - no atom count demerits\n"
  $stderr << " -smarts <s>      optional smarts to reject\n" if $expert
  $stderr << " -rej <q>         optional query file to reject\n" if $expert
  $stderr << " -relaxed         relaxed rules: 7-50 heavy atoms, 160 demerit cutoff\n"
  $stderr << " -nodemerit       hard rejections only, do not apply any demerits\n" if $expert
  $stderr << " -S <fname>       write output to <fname> rather than stdout\n" if $expert
  $stderr << " -B <stem>        output name stem for rejected molecules\n" if $expert
  $stderr << " -log <stem>      name stem for log files\n" if $expert
  $stderr << " -bindir <dir>    directory for executables\n" if $expert
  $stderr << " -tp...-tp        options passed directly to mc_first_pass\n" if $expert
  $stderr << " -iwd...-iwd      options passed directly to iwdemerit\n" if $expert
  $stderr << " -odm <name>      omit demerit with file name <name>\n" if $expert
  $stderr << " -edm <fname>     extra demerits to be applied - query file format only\n" if $expert
  $stderr << " -dmrt <fname>    extra demerits to be applied\n" if $expert
  $stderr << " -dcf  <fname>    demerit control file (the -C option to iwdemerit)\n" if $expert
  $stderr << " -q <dir>         directory for queries\n" if $expert
  $stderr << " -okiso           allow isotopic atoms to pass through\n";
  $stderr << " -label           place isotopic labels on the matched atoms for rejected molecules (bad1.smi and bad2.smi)\n" if $expert
  $stderr << " -symm <bonds>    discard symmetric molecules where two symmetric atoms > <bonds> apart\n" if $expert
  $stderr << " -nophosphorus    reject all phosphorus containing molecules\n"
  $stderr << " -noapdm          do not append demerit reasons\n"
  $stderr << " -i <type>        input type\n" if $expert
  $stderr << " -expert          more options\n" unless $expert;
  $stderr << " -v               verbose output\n"
  exit(rc)
end

cl = IWCmdline.new("-v-noapdm-i=s-expert-b=fraction-B=s-q=dir-log=s-tp=close-iwd=close-bindir=dir-smarts=s-rej=s-c=ipos-Cs=ipos-Ch=ipos-C=ipos-okiso-odm=s-edm=sfile-relaxed-nodemerit-S=s-dcf=sfile-nobadfiles-symm=ipos-nophosphorus-label")

if cl.unrecognised_options_encountered()
  $stderr << "Unrecognised options encountered\n"
  usage(1)
end

verbose = cl.option_present('v')

if cl.option_present('expert')
  $expert = true
end

if ARGV.empty?
  $stderr << "Insufficient arguments\n"
  usage(2)
end

append_demerit_reason = ! cl.option_present('noapdm')   # convoluted inverse logic

stop_afer_completing_step = 3    # no longer optional

input_type = ""
if cl.option_present('i')
  input_type = '-i ' << cl.value('i')
end

# The ring bond ratio used by mc_first_pass

ring_bond_ratio = -1
if cl.option_present('b')
  ring_bond_ratio = cl.value('b')
end

mc_first_pass_options = ""
if cl.option_present('tp')
  mc_first_pass_options = cl.value('tp')
end

if cl.option_present('nophosphorus')
  mc_first_pass_options << " -n P"
end

unless cl.option_present('okiso')
  mc_first_pass_options << ' -I 0'
end

mc_first_pass_options << ' -A I -A ipp'

bindir = Bindir.new(ianhome)

cl.values('bindir').each do |d|
  bindir.add_dir(d)
end

iwdemerit     = bindir.find_executable('iwdemerit')
mc_first_pass = bindir.find_executable('mc_first_pass')
tsubstructure = bindir.find_executable('tsubstructure')

query_dir = "#{ianhome}/queries";

if cl.option_present('q')
  query_dir = cl.value('q')
end

# The default query files

query_file = Array.new

query_file[1] = 'reject1'
query_file[2] = 'reject2'
query_file[3] = 'demerits'

# The default stem of the rejected files

bad_stem = false

if cl.option_present('B')
  bad_stem = cl.value('B')
elsif cl.option_present('nobadfiles')
  true
else
  bad_stem = 'bad'
end

# The default stem for log files

logfilestem = 'ok';

if cl.option_present('log')
  logfilestem = cl.value('log')
end

optional_queries = ""

cl.values('rej').each do |q|
  optional_queries << " -q #{q}"
end

cl.values('smarts').each do |s|
  optional_queries << " -s '#{s}'"
end

extra_iwdemerit_options = ""

if cl.option_present('relaxed')
  $default_soft_upper_atom_count_cutoff = 26
  $default_hard_upper_atom_count_cutoff = 50
  extra_iwdemerit_options << " -f 160"
end

if cl.option_present('nodemerit')
  extra_iwdemerit_options << " -r"
  $default_soft_upper_atom_count_cutoff = $default_hard_upper_atom_count_cutoff - 1
end

if cl.option_present('iwd')
  extra_iwdemerit_options << " " << cl.value('iwd')
end

if cl.option_present('symm')
  extra_iwdemerit_options << " -s " << cl.value('symm').to_s
end

charge_assigner = "#{ianhome}/charge_assigner/queries"

unless FileTest.size?(charge_assigner)
  $stderr << "Charge assigner not available, skipping\n"
else
  extra_iwdemerit_options << " -N F:#{charge_assigner}"
end

lower_atom_count_cutoff = $default_lower_atom_count_cutoff
if cl.option_present('c')
  lower_atom_count_cutoff = cl.value('c')
end

soft_upper_atom_count_cutoff = $default_soft_upper_atom_count_cutoff;
if cl.option_present('Cs')
  soft_upper_atom_count_cutoff = cl.value('Cs')
end

hard_upper_atom_count_cutoff = $default_hard_upper_atom_count_cutoff;
if cl.option_present('Ch')
  hard_upper_atom_count_cutoff = cl.value('Ch')
  unless cl.option_present('Cs')
    soft_upper_atom_count_cutoff = hard_upper_atom_count_cutoff - 1
  end
end

if cl.option_present('C')
  hard_upper_atom_count_cutoff = cl.value('C') + 1
  soft_upper_atom_count_cutoff = hard_upper_atom_count_cutoff - 1
end

if hard_upper_atom_count_cutoff < soft_upper_atom_count_cutoff
  hard_upper_atom_count_cutoff = soft_upper_atom_count_cutoff + 1
end

unless FileTest.directory?(query_dir)
  $stderr << "Cannot continue, query dir '#{query_dir}' invalid\n"
  exit(3)
end

raise "Query file '#{query_dir}/#{query_file[1]}' missing or inaccessible" unless (FileTest.size?("#{query_dir}/#{query_file[1]}") && FileTest.readable?("#{query_dir}/#{query_file[1]}"))
raise "Query file '#{query_dir}/#{query_file[2]}' missing or inaccessible" unless (FileTest.size?("#{query_dir}/#{query_file[2]}") && FileTest.readable?("#{query_dir}/#{query_file[2]}"))
raise "Query file '#{query_dir}/#{query_file[3]}' missing or inaccessible" unless (FileTest.size?("#{query_dir}/#{query_file[3]}") && FileTest.readable?("#{query_dir}/#{query_file[3]}"))

$stderr << "Queries from '#{query_dir}'\n" if (verbose)

files_to_be_deleted = Array.new

query_file3 = "#{query_dir}/#{query_file[3]}"

# I want to have the odm option behave as either a regular expression
# or as exact matches.
# Note that I don't check against multiple RX= directives

class Odm
  def initialize(o)
    @rx = false
    @hash = Hash.new

    o.each do |d|
      if ('RX=' == d[0,3])
        @rx = Regexp.new(d[3, d.size - 3], Regexp::IGNORECASE)
      else
        @hash[d] = true
      end
    end
  end

  def match(d)
    if (@hash.has_key?(d))
      return true
    end

    if (@rx)
      return @rx.match(d)
    end

    return false
  end
end

if cl.option_present('odm')

  tmpdir = "."
  if cl.option_present('tmpdir')
    tmpdir = cl.value('tmpdir')
  end

  old_demerits = File.open("#{query_dir}/#{query_file[3]}", mode='r')
  raise "Cannot open original demerit file '#{query_dir}/#{query_file[3]}'" unless (old_demerits)

  temporary_demerit_file = "#{tmpdir}/demerits" << Process.pid.to_s
  new_demerits = File.open(temporary_demerit_file, mode='w')
  raise "Cannot open temporary demerit file '#{temporary_demerit_file}'" unless (new_demerits)

  odm = Odm.new(cl.values('odm'))

  demerit_rx = Regexp.new("^(\\S+)\.qry")

  items_discarded = 0

  old_demerits.each do |line|
    m = demerit_rx.match(line)

    next unless (m)

    stem = m[1]

    if odm.match(stem)
      items_discarded += 1
    else
      new_demerits << "#{query_dir}/#{stem}.qry\n"
    end
  end

  if 0 == items_discarded
    $stderr << "Warning, no demerits discarded\n"
  elsif verbose
    $stderr << "Discarded #{items_discarded} demerits\n"
  end

  new_demerits.close

  query_file3 = "#{temporary_demerit_file}"

  files_to_be_deleted.push(query_file3)
end

iwdemerit_optional_control_file = false

if cl.option_present('dcf')
  iwdemerit_optional_control_file = cl.value('dcf')
end

if cl.option_present('edm')
  additional_demerits = cl.value('edm')
end

cmd = "#{mc_first_pass} ";

cmd << " -b #{ring_bond_ratio}" if ring_bond_ratio >= 0.0

cmd << " #{mc_first_pass_options}" if mc_first_pass_options.length > 0

cmd << " #{input_type} " if input_type.length > 0

cmd << " -c #{lower_atom_count_cutoff} -C #{hard_upper_atom_count_cutoff} -E autocreate -o smi -V -g all -g ltltr -i ICTE "
cmd << "-L #{bad_stem}0 -K TP1 " if bad_stem
cmd << "-a -S - #{ARGV.join(' ')} 2> #{logfilestem}0.log "

# Options passed to both tsubstructure invocations
tsub_common_opts = '-A D -E autocreate -b -u -i smi -o smi'

if cl.option_present('label')
  tsub_common_opts << ' -j 1'
end

if stop_afer_completing_step >= 1
  cmd << "| #{tsubstructure} #{tsub_common_opts} "
  cmd << "-m #{bad_stem}1 -m QDT " if bad_stem
  cmd << "-n - -q F:#{query_dir}/#{query_file[1]} "

  cmd << optional_queries if optional_queries.length > 0

  cmd << " - 2> #{logfilestem}1.log ";

  if stop_afer_completing_step >= 2
    cmd << "| #{tsubstructure} #{tsub_common_opts} "
    cmd << "-m #{bad_stem}2 -m QDT " if bad_stem
    cmd << "-n - -q F:#{query_dir}/#{query_file[2]} - 2> #{logfilestem}2.log ";
    if stop_afer_completing_step >= 3
      cmd << " | #{iwdemerit} -x #{extra_iwdemerit_options} -E autocreate -A D -i smi -o smi -q F:#{query_file3} "
      cmd << "-R #{bad_stem}3 " if bad_stem
      cmd << "-G - -c smax=#{soft_upper_atom_count_cutoff} -c hmax=#{hard_upper_atom_count_cutoff} "
      cmd << "-q F:#{additional_demerits} " if additional_demerits
      cmd << "-C #{iwdemerit_optional_control_file} " if iwdemerit_optional_control_file
      cmd << "-t " if append_demerit_reason
      cmd << "- 2> #{logfilestem}3.log "
    end
  end
end

if cl.option_present('S')
  s = cl.value('S')
  cmd << " > #{s}"
end

$stderr << "Command is '#{cmd}'\n" if verbose

system(cmd)

files_to_be_deleted.each do |f|
  File.unlink(f) if FileTest.exists?(f)
end

File: Lilly_Medchem_Rules.py
#!/usr/bin/env python3
"""Python driver for the Lilly Medchem Rules pipeline.

This is intentionally a conservative translation of Lilly_Medchem_Rules.rb.
It preserves the overall shell pipeline construction and calls the existing
LillyMol executables rather than reimplementing chemistry functionality.

Important compatibility choice:
  * Wrapper-owned paths/files are shell-quoted.
  * Pass-through fragments from -tp and -iwd are appended unquoted, so a user can
    write -iwd 'v1 v2' and have v1 and v2 arrive as separate downstream tokens.
"""

from __future__ import annotations

import argparse
import os
import re
import shlex
import subprocess
import sys
from pathlib import Path


DEFAULT_LOWER_ATOM_COUNT_CUTOFF = 7
DEFAULT_SOFT_UPPER_ATOM_COUNT_CUTOFF = 25
DEFAULT_HARD_UPPER_ATOM_COUNT_CUTOFF = 40


class Odm:
    """Matches demerit names either by exact name or by one RX= regex."""

    def __init__(self, omit_specs: list[str]):
        self._names: set[str] = set()
        self._rx: re.Pattern[str] | None = None

        for spec in omit_specs:
            if spec.startswith("RX="):
                self._rx = re.compile(spec[3:], re.IGNORECASE)
            else:
                self._names.add(spec)

    def match(self, name: str) -> bool:
        if name in self._names:
            return True
        if self._rx is not None and self._rx.search(name):
            return True
        return False


def quote(value: str | Path) -> str:
    return shlex.quote(str(value))


def readable_nonempty_file(path: Path) -> bool:
    return path.is_file() and path.stat().st_size > 0 and os.access(path, os.R_OK)


def executable_file(path: Path) -> bool:
    return path.is_file() and os.access(path, os.X_OK)


def find_executable(name: str, search_dirs: list[Path]) -> str:
    """Find an executable, falling back to PATH lookup by command name."""
    for directory in search_dirs:
        candidate = directory / name
        if executable_file(candidate):
            return quote(candidate)

    return name


def script_home() -> Path:
    return Path(__file__).resolve().parent


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Run the Lilly medchem rejection/demerit rules."
    )

    parser.add_argument("-v", action="store_true", help="Verbose output")
    parser.add_argument("-expert", action="store_true", help="Accepted for compatibility")
    parser.add_argument("-noapdm", action="store_true", help="Do not append demerit reasons")
    parser.add_argument("-i", metavar="TYPE", help="Input type")
    parser.add_argument("-b", type=float, metavar="FRACTION", help="Ring bond ratio")
    parser.add_argument("-B", metavar="STEM", help="Output name stem for rejected molecules")
    parser.add_argument("-q", metavar="DIR", help="Directory for query files")
    parser.add_argument("-log", metavar="STEM", help="Name stem for log files")

    # Pass-through fragments. These are intentionally appended unquoted.
    parser.add_argument(
        "-tp",
        action="append",
        default=[],
        metavar="OPTS",
        help="Options passed directly to mc_first_pass. Quote multiple shell tokens.",
    )
    parser.add_argument(
        "-iwd",
        action="append",
        default=[],
        metavar="OPTS",
        help="Options passed directly to iwdemerit. Quote multiple shell tokens.",
    )

    parser.add_argument(
        "-bindir",
        action="append",
        default=[],
        metavar="DIR",
        help="Directory containing executables. Can be repeated.",
    )
    parser.add_argument(
        "-smarts",
        action="append",
        default=[],
        metavar="SMARTS",
        help="Optional SMARTS to reject. Can be repeated.",
    )
    parser.add_argument(
        "-rej",
        action="append",
        default=[],
        metavar="QUERY",
        help="Optional query file to reject. Can be repeated.",
    )

    parser.add_argument("-c", type=int, metavar="N", help="Lower atom count cutoff")
    parser.add_argument("-Cs", type=int, metavar="N", help="Soft upper atom count cutoff")
    parser.add_argument("-Ch", type=int, metavar="N", help="Hard upper atom count cutoff")
    parser.add_argument("-C", type=int, metavar="N", help="Upper atom count limit; no atom count demerits")

    parser.add_argument("-okiso", action="store_true", help="Allow isotopic atoms")
    parser.add_argument(
        "-odm",
        action="append",
        default=[],
        metavar="NAME_OR_RX",
        help="Omit demerit by exact stem or RX=regex. Can be repeated.",
    )
    parser.add_argument("-edm", metavar="FILE", help="Extra demerits query file")
    parser.add_argument("-relaxed", action="store_true", help="Relaxed rules")
    parser.add_argument("-nodemerit", action="store_true", help="Hard rejections only")
    parser.add_argument("-S", metavar="FILE", help="Write output to file rather than stdout")
    parser.add_argument("-dcf", metavar="FILE", help="Demerit control file for iwdemerit")
    parser.add_argument("-nobadfiles", action="store_true", help="Do not create bad*.smi files")
    parser.add_argument(
        "-symm",
        type=int,
        metavar="N",
        help="Discard symmetric molecules where two symmetric atoms are more than N apart",
    )
    parser.add_argument(
        "-nophosphorus",
        action="store_true",
        help="Reject all phosphorus-containing molecules",
    )
    parser.add_argument(
        "-label",
        action="store_true",
        help="Place isotopic labels on matched atoms for rejected molecules",
    )
    parser.add_argument(
        "-tabular",
        action="store_true",
        help="Tabular output 'smiles id demerit' with non demerited molecules getting '0'"
    )

    # This fixes the Ruby bug where tmpdir was checked but not declared.
    parser.add_argument(
        "-tmpdir",
        default=".",
        metavar="DIR",
        help="Directory for temporary files created by -odm",
    )

    parser.add_argument("inputs", nargs="+", help="Input molecule files")
    return parser


def validate_queries(query_dir: Path) -> None:
    for name in ("reject1", "reject2", "demerits"):
        path = query_dir / name
        if not readable_nonempty_file(path):
            raise RuntimeError(f"Query file '{path}' missing or inaccessible")


def build_odm_demerit_file(
    query_dir: Path, tmpdir: Path, omit_specs: list[str], verbose: bool
) -> tuple[Path, list[Path]]:
    """Build a temporary demerits file after dropping selected demerits."""
    old_demerits = query_dir / "demerits"
    if not readable_nonempty_file(old_demerits):
        raise RuntimeError(f"Cannot open original demerit file '{old_demerits}'")

    tmpdir.mkdir(parents=True, exist_ok=True)
    temporary_demerit_file = tmpdir / f"demerits{os.getpid()}"

    odm = Odm(omit_specs)
    demerit_rx = re.compile(r"^(\S+)\.qry")
    items_discarded = 0

    with old_demerits.open("r", encoding="utf-8") as inp, temporary_demerit_file.open(
        "w", encoding="utf-8"
    ) as out:
        for line in inp:
            match = demerit_rx.match(line)
            if not match:
                continue
            stem = match.group(1)
            if odm.match(stem):
                items_discarded += 1
            else:
                out.write(f"{query_dir / (stem + '.qry')}\n")

    if items_discarded == 0:
        print("Warning, no demerits discarded", file=sys.stderr)
    elif verbose:
        print(f"Discarded {items_discarded} demerits", file=sys.stderr)

    return temporary_demerit_file, [temporary_demerit_file]


def build_command(args: argparse.Namespace) -> tuple[str, list[Path]]:
    home = script_home()

    lower_atom_count_cutoff = DEFAULT_LOWER_ATOM_COUNT_CUTOFF
    soft_upper_atom_count_cutoff = DEFAULT_SOFT_UPPER_ATOM_COUNT_CUTOFF
    hard_upper_atom_count_cutoff = DEFAULT_HARD_UPPER_ATOM_COUNT_CUTOFF

    append_demerit_reason = not args.noapdm

    input_type = ""
    if args.i:
        input_type = f"-i {args.i}"

    ring_bond_ratio = -1.0
    if args.b is not None:
        ring_bond_ratio = args.b

    mc_first_pass_options = ""
    if args.tp:
        mc_first_pass_options = " ".join(args.tp)  # pass-through, intentionally unquoted

    if args.nophosphorus:
        mc_first_pass_options += " -n P"

    if not args.okiso:
        mc_first_pass_options += " -I 0"

    mc_first_pass_options += " -A I -A ipp"

    search_dirs = [home]
    search_dirs.extend(Path(d) for d in args.bindir)
    search_dirs.extend(
        [
            home / "bin" / "Linux",
            home / "bin",
            home / "build",
        ]
    )

    iwdemerit = find_executable("iwdemerit", search_dirs)
    mc_first_pass = find_executable("mc_first_pass", search_dirs)
    tsubstructure = find_executable("tsubstructure", search_dirs)

    query_dir = Path(args.q) if args.q else home / "queries"
    if not query_dir.is_dir():
        raise RuntimeError(f"Cannot continue, query dir '{query_dir}' invalid")
    validate_queries(query_dir)

    if args.v:
        print(f"Queries from '{query_dir}'", file=sys.stderr)

    if args.B:
        bad_stem: str | None = args.B
    elif args.nobadfiles:
        bad_stem = None
    else:
        bad_stem = "bad"

    logfilestem = args.log if args.log else "ok"

    optional_queries = ""
    for query in args.rej:
        optional_queries += f" -q {quote(query)}"
    for smarts in args.smarts:
        # Match Ruby behavior of surrounding SMARTS with shell quotes.
        optional_queries += f" -s {quote(smarts)}"

    extra_iwdemerit_options = ""
    if args.relaxed:
        soft_upper_atom_count_cutoff = 26
        hard_upper_atom_count_cutoff = 50
        extra_iwdemerit_options += " -f 160"

    if args.nodemerit:
        extra_iwdemerit_options += " -r"
        soft_upper_atom_count_cutoff = hard_upper_atom_count_cutoff - 1

    if args.iwd:
        extra_iwdemerit_options += " " + " ".join(args.iwd)  # pass-through, intentionally unquoted

    if args.symm is not None:
        extra_iwdemerit_options += f" -s {args.symm}"

    charge_assigner = home / "charge_assigner" / "queries"
    if not readable_nonempty_file(charge_assigner):
        print("Charge assigner not available, skipping", file=sys.stderr)
    else:
        extra_iwdemerit_options += f" -N F:{quote(charge_assigner)}"

    if args.c is not None:
        lower_atom_count_cutoff = args.c

    if args.Cs is not None:
        soft_upper_atom_count_cutoff = args.Cs

    if args.Ch is not None:
        hard_upper_atom_count_cutoff = args.Ch
        if args.Cs is None:
            soft_upper_atom_count_cutoff = hard_upper_atom_count_cutoff - 1

    if args.C is not None:
        hard_upper_atom_count_cutoff = args.C + 1
        soft_upper_atom_count_cutoff = hard_upper_atom_count_cutoff - 1

    if hard_upper_atom_count_cutoff < soft_upper_atom_count_cutoff:
        hard_upper_atom_count_cutoff = soft_upper_atom_count_cutoff + 1

    files_to_delete: list[Path] = []
    query_file3: Path = query_dir / "demerits"

    if args.odm:
        query_file3, files_to_delete = build_odm_demerit_file(
            query_dir=query_dir,
            tmpdir=Path(args.tmpdir),
            omit_specs=args.odm,
            verbose=args.v,
        )

    iwdemerit_optional_control_file = args.dcf
    additional_demerits = args.edm
    tabular_output = args.tabular

    quoted_inputs = " ".join(quote(x) for x in args.inputs)

    cmd = f"{mc_first_pass} "
    if ring_bond_ratio >= 0.0:
        cmd += f" -b {ring_bond_ratio}"
    if mc_first_pass_options:
        cmd += f" {mc_first_pass_options}"
    if input_type:
        cmd += f" {input_type} "

    cmd += (
        f" -c {lower_atom_count_cutoff} -C {hard_upper_atom_count_cutoff} "
        "-E autocreate -o smi -V -g all -g ltltr -i ICTE "
    )
    if bad_stem:
        cmd += f"-L {quote(bad_stem + '0')} -K TP1 "
    cmd += f"-a -S - {quoted_inputs} 2> {quote(logfilestem + '0.log')} "

    tsub_common_opts = "-A D -E autocreate -b -u -i smi -o smi"
    if args.label:
        tsub_common_opts += " -j 1"

    cmd += f"| {tsubstructure} {tsub_common_opts} "
    if bad_stem:
        cmd += f"-m {quote(bad_stem + '1')} -m QDT "
    cmd += f"-n - -q F:{quote(query_dir / 'reject1')} "
    if optional_queries:
        cmd += optional_queries
    cmd += f" - 2> {quote(logfilestem + '1.log')} "

    cmd += f"| {tsubstructure} {tsub_common_opts} "
    if bad_stem:
        cmd += f"-m {quote(bad_stem + '2')} -m QDT "
    cmd += f"-n - -q F:{quote(query_dir / 'reject2')} - 2> {quote(logfilestem + '2.log')} "

    cmd += f" | {iwdemerit} -x {extra_iwdemerit_options} "
    cmd += f"-E autocreate -A D -i smi -o smi -q F:{quote(query_file3)} "
    if bad_stem:
        cmd += f"-R {quote(bad_stem + '3')} "
    cmd += (
        f"-G - -c smax={soft_upper_atom_count_cutoff} "
        f"-c hmax={hard_upper_atom_count_cutoff} "
    )
    if additional_demerits:
        cmd += f"-q F:{quote(additional_demerits)} "
    if tabular_output:
      cmd += "-W tabular "
    if iwdemerit_optional_control_file:
        cmd += f"-C {quote(iwdemerit_optional_control_file)} "
    if append_demerit_reason:
        cmd += "-t "
    cmd += f"- 2> {quote(logfilestem + '3.log')} "

    if args.S:
        cmd += f" > {quote(args.S)}"

    return cmd, files_to_delete


def main(argv: list[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)

    try:
        cmd, files_to_delete = build_command(args)
    except Exception as exc:  # Keep behavior simple for command-line users.
        print(str(exc), file=sys.stderr)
        return 1

    if args.v:
        print(f"Command is '{cmd}'", file=sys.stderr)

    try:
        completed = subprocess.run(cmd, shell=True, check=False)
        return_code = completed.returncode
    finally:
        for path in files_to_delete:
            try:
                path.unlink()
            except FileNotFoundError:
                pass

    return return_code


if __name__ == "__main__":
    sys.exit(main())
